orchagent/pfcactionhandler.cpp

```diff
@@ -0,0 +1,466 @@
+
+#define PFC_WD_QUEUE_STATS_DEADLOCK_DETECTED "PFC_WD_QUEUE_STATS_DEADLOCK_DETECTED"
+#define PFC_WD_QUEUE_STATS_DEADLOCK_RESTORED "PFC_WD_QUEUE_STATS_DEADLOCK_RESTORED"
+
+#define SAI_QUEUE_STAT_PACKETS_STR         "SAI_QUEUE_STAT_PACKETS"
+#define SAI_QUEUE_STAT_DROPPED_PACKETS_STR "SAI_QUEUE_STAT_DROPPED_PACKETS"
+
+extern sai_object_id_t gSwitchId;
+extern sai_port_api_t *sai_port_api;
+extern sai_queue_api_t *sai_queue_api;
+extern sai_buffer_api_t *sai_buffer_api;
+
+PfcWdActionHandler::PfcWdActionHandler(sai_object_id_t port, sai_object_id_t queue,
+        uint8_t queueId, shared_ptr<Table> countersTable):
+    m_port(port),
+    m_queue(queue),
+    m_queueId(queueId),
+    m_countersTable(countersTable)
+{
+    SWSS_LOG_ENTER();
+
+    SWSS_LOG_NOTICE(
+            "PFC Watchdog detected PFC storm on queue 0x%lx port 0x%lx",
+            m_queue,
+            m_port);
+    
+    m_stats = getQueueStats(m_countersTable, sai_serialize_object_id(m_queue));
+    m_stats.detectCount++;
+    m_stats.operational = false;
+
+    updateWdCounters(sai_serialize_object_id(m_queue), m_stats);
+}
+
+PfcWdActionHandler::~PfcWdActionHandler(void)
+{
+    SWSS_LOG_ENTER();
+
+    auto finalStats = getQueueStats(m_countersTable, sai_serialize_object_id(m_queue));
+
+    SWSS_LOG_NOTICE(
+            "Queue 0x%lx port 0x%lx restored from PFC storm. Tx packets: %lu. Dropped packets: %lu",
+            m_queue,
+            m_port,
+            finalStats.txPkt - m_stats.txPkt,
+            finalStats.txDropPkt - m_stats.txDropPkt);
+
+    finalStats.restoreCount++;
+    finalStats.operational = true;
+
+    updateWdCounters(sai_serialize_object_id(m_queue), finalStats);
+}
+
+PfcWdActionHandler::PfcWdQueueStats PfcWdActionHandler::getQueueStats(shared_ptr<Table> countersTable, const string &queueIdStr)
+{
+    SWSS_LOG_ENTER();
+
+    PfcWdQueueStats stats;
+    vector<FieldValueTuple> fieldValues;
+
+    if (!countersTable->get(queueIdStr, fieldValues))
+    {
+        return move(stats);
+    }
+
+    for (const auto& fv : fieldValues)
+    {
+        const auto field = fvField(fv);
+        const auto value = fvValue(fv);
+
+        if (field == PFC_WD_QUEUE_STATS_DEADLOCK_DETECTED)
+        {
+            stats.detectCount = stoul(value);
+        }
+        else if (field == PFC_WD_QUEUE_STATS_DEADLOCK_RESTORED)
+        {
+            stats.restoreCount = stoul(value);
+        }
+        else if (field == PFC_WD_QUEUE_STATUS)
+        {
+            stats.operational = value == PFC_WD_QUEUE_STATUS_OPERATIONAL ? true : false;
+        }
+        else if (field == SAI_QUEUE_STAT_PACKETS_STR)
+        {
+            stats.txPkt = stoul(value);
+        }
+        else if (field == SAI_QUEUE_STAT_DROPPED_PACKETS_STR)
+        {
+            stats.txDropPkt = stoul(value);
+        }
+    }
+
+    return move(stats);
+}
+
+void PfcWdActionHandler::initWdCounters(shared_ptr<Table> countersTable, const string &queueIdStr)
```

Comment:
Do we need this function? Looks like it gets the counter information and set them back?

Reply:
Yes. It's for user's convenience. It fills counters in DB with zeros. if they are absent.